### ISSUE TITLE
Add shape_idx to mouse_entered signal of CollisionObject2D

### ISSUE THREAD

**Issue by sygi** [godotengine/godot-proposals#2487]
### Describe the project you are working on
A point-and-click adventure game.

### Describe the problem or limitation you are having in your project
I am using CollisionObject2D (concretely its subclass Area2D) for changing the shape of the mouse pointer depending on the underlying object (described by a Shape2D). The CollisionObject2D has 3 [signals](https://docs.godotengine.org/en/latest/classes/class_collisionobject2d.html#signals) related to mouse movement:
```
- input_event ( Node viewport, InputEvent event, int shape_idx )
- mouse_entered ( )
- mouse_exited ( )
```
I would like to, on getting a `mouse_entered` signal, change the mouse based on the particular underlying Shape, but this signal is not passed `shape_idx`. Note a similar issue  #1574

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
I would like to extend `mouse_entered/mouse_exited` signals to pass `shape_idx` as an argument.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
It should be possible to pass the resulting shape somewhere around [here](https://github.com/godotengine/godot/blob/08ca4184f407419f4265bdc0add21ac2e78bef7e/scene/main/viewport.cpp#L730).

### If this enhancement will not be used often, can it be worked around with a few lines of script?
Yee-es, but I believe enhancing the existing signal is a more natural and less error-prone way to do so. Getting this extra argument is not a big cost to pay.

### Is there a reason why this should be core and not an add-on in the asset library?
It changes the code API.

**Comment by sygi** [comment]
I tried it out and indeed passing the `res[i].shape` works for `mouse_entered` (see [code](https://github.com/godotengine/godot/pull/47395)). However, I realized a following problem:
if there are multiple neighbouring (or intersecting) shapes within a single `CollisionObject2D`, we currently don't trigger `mouse_entered/exited` when moving between them. Because of that, if a user enters shape `A`, then moves to shape `B` and then exits, triggering `mouse_exited(B)` may be confusing (and will require some more coding, as we currently don't store the particular shape id, only the id of the `CollisionObject`), as someone may be trying to match `exit` events with `entered` ones.

I see three potential solutions:
1. Trigger `mouse_entered/exited` every time we move between different shapes, even if they are within a single `CollisionObject2D` (changing the semantics of the signal).
2. Send the `shape_idx` only when entering the CollisionObject2D which will be enough to solve my problem described but look somewhat inconsistent.
3. Track which shape id was entered/left, but only send the signal upon moving out of the `CollisionObject2D` (the behavior marked by me as confusing above).

What do you think?

**Comment by pouleyKetchoupp** [comment]
It does seem useful to get the shape information, but maybe instead we could add 2 new signals `mouse_shape_entered(shape_idx)` and `mouse_shape_exited(shape_idx)` which would follow the semantics from your point 1? 

This way it wouldn't break compatibility and what to expect would be perfectly clear to users.

Looking at the code, it should be relatively easy to add a shape index in physics_2d_mouseover in order to handle both signals at the same time.

**Comment by sygi** [comment]
Thanks for the comment, that makes sense.

Do you have a preference over a single signal:
`mouse_shape_changed(shape_idx)` with `-1` aligning with `mouse_exited`?

**Comment by pouleyKetchoupp** [comment]
> Do you have a preference over a single signal:
mouse_shape_changed(shape_idx) with -1 aligning with mouse_exited?

I would rather go for a separate set of signals for shapes, it will be easier to document them clearly.

**Comment by sygi** [comment]
I added the two signals in the PR and changed `physics_2d_mouseover` to `Map<ObjectID, Pair<uint64_t, int>>`.

I have realized though that right now, if there are intersecting `Shape2D`s, the two signals will be repeatedly called. I don't think there's a way to fix it without introducing a separate map for `collider + shape -> frame` (or making `physics_2d_mouseover` `collider -> shape -> frame` and iterating over all previous shapes): we need separate information for "have this collider changed" vs "have this shape changed".

Sorry if this is requires too detailed guidance: I want to better understand what makes sense to you :)
